### PR TITLE
refactor(NODE-3291)!: Standardize error representation in the driver

### PR DESCRIPTION
**Pull request by dariakp**
## Description

NODE-3291

BREAKING: Previously thrown `TypeError` instances have been converted to `MongoDriverError` instances and `MongoParseError` instances (which extend from the same class). Error names have changed to be more specific.

All thrown errors across the codebase have been converted to extend from a specific subclass of `MongoError`: 
- `MongoServerError` when wrapping errors returned from the server on specific queries, 
- `MongoDriverError` when generated by the driver, 
- `MongoNetworkError` when associated with a network connection failure, 
- `MongoSystemError` to represent errors associated with general system configuration.

### CODE REVIEW

File: src/bulk/common.ts [hunk 1]
```diff
@@ -649,16 +655,17 @@ function handleMongoWriteConcernError(
 ) {
   mergeBatchResults(batch, bulkResult, undefined, err.result);
 
+  // TODO: make more reasonable
```
Comment (dariakp) [inline src/bulk/common.ts:658]: This needs to be created as a ticket in the error improvements epic, currently captured in doc

File: src/gridfs-stream/download.ts [hunk 2]
```diff
@@ -195,7 +195,9 @@ export class GridFSBucketReadStream extends Readable {
 
 function throwIfInitialized(stream: GridFSBucketReadStream): void {
   if (stream.s.init) {
-    throw new Error('You cannot change options after the stream has entered flowing mode!');
+    throw new MongoDriverError(
+      'You cannot change options after the stream has entered flowing mode!'
```
Comment (nbbeeken) [inline src/gridfs-stream/download.ts:199]: clean up request, Can we rephrase this to something like `Options cannot be changed after stream is initialized`?

File: src/gridfs-stream/download.ts [hunk 3]
```diff
@@ -313,8 +315,8 @@ function init(stream: GridFSBucketReadStream): void {
         ? stream.s.filter._id.toString()
         : stream.s.filter.filename;
       const errmsg = 'FileNotFound: file ' + identifier + ' was not found';
-      const err = new Error(errmsg);
-      (err as any).code = 'ENOENT';
+      const err = new MongoDriverError(errmsg);
+      err.code = 'ENOENT';
```
Comment (nbbeeken) [inline src/gridfs-stream/download.ts:319]: Just commenting so we remember to make a ticket for this.

File: src/gridfs-stream/download.ts [hunk 4]
```diff
@@ -237,28 +239,28 @@ function doRead(stream: GridFSBucketReadStream): void {
...
     let buf = Buffer.isBuffer(doc.data) ? doc.data : doc.data.buffer;
 
     if (buf.byteLength !== expectedLength) {
       if (bytesRemaining <= 0) {
         errmsg = 'ExtraChunk: Got unexpected n: ' + doc.n;
-        return __handleError(stream, new Error(errmsg));
+        return __handleError(stream, new MongoDriverError(errmsg));
       }
 
       errmsg =
```
Comment (nbbeeken) [inline src/gridfs-stream/download.ts:258]: There's a lot of concat strings here that could template strings, its understandable if you wouldn't like to add on this PR further with some code clean up, but figure I'd suggest we do it here anyway. Feel free to resolve without changes.

File: src/operations/eval.ts [hunk 5]
```diff
@@ -68,7 +68,7 @@ export class EvalOperation extends CommandOperation<Document> {
       }
 
       if (result) {
-        callback(MongoError.create({ message: `eval failed: ${result.errmsg}`, driver: true }));
+        callback(new MongoServerError({ message: `eval failed: ${result.errmsg}`, driver: true }));
```
Comment (nbbeeken) [inline src/operations/eval.ts:71]:
```suggestion
callback(new MongoServerError(`eval failed: ${result.errmsg}`));
```

The `driver: true` flag is legacy from when mongodb-core was a seperate package. We don't need to keep this.

File: src/collection.ts [hunk 6]
```diff
@@ -1400,7 +1400,7 @@ export class Collection<TSchema extends Document = Document> {
     if ('function' === typeof options) (callback = options), (options = {});
     // Out must always be defined (make sure we don't break weirdly on pre 1.8+ servers)
```
Comment (nbbeeken) [inline src/collection.ts:1401]: This seems like something we should investigate as we only support 2.6+ servers. Commenting to remember to ticket.

File: src/error.ts [hunk 7]
```diff
@@ -383,14 +399,7 @@ export function isSDAMUnrecoverableError(error: MongoError): boolean {
...
 
-  if (typeof error.code !== 'undefined') {
-    return isRecoveringError(error) || isNotMasterError(error);
-  }
-  if (isRecoveringError(error) || isNotMasterError(error)) {
-    return true;
-  }
-
-  return false;
+  return isRecoveringError(error) || isNotMasterError(error);
```
Comment (nbbeeken) [inline src/error.ts:402]: Nice clean up :clap:

File: src/sessions.ts [hunk 8]
```diff
@@ -761,22 +771,22 @@ function applySession(
   session: ClientSession,
   command: Document,
   options?: CommandOptions
-): MongoError | undefined {
+): MongoDriverError | undefined {
   // TODO: merge this with `assertAlive`, did not want to throw a try/catch here
   if (session.hasEnded) {
-    return new MongoError('Attemped to use a session that has ended');
+    return new MongoDriverError('Attemped to use a session that has ended');
```
Comment (nbbeeken) [inline src/sessions.ts:777]: clean up request: `Attemped` -> `Attempted`

**Review by nbbeeken — CHANGES_REQUESTED** [review]
I just have a number of clean up suggestions which some could be left to the follow up work that's coming up soon, I leave that to you to decide. There's on instance where a MongoError.create call's arguments were preserved where they need not be. Beyond that this is looking very ready to merge.

File: src/sessions.ts (hunk 8 above)
Comment (dariakp) [reply src/sessions.ts:777]: Nice catch, I wasn't even looking at the messages themselves

File: src/cmap/wire_protocol/compression.ts [hunk 9]
```diff
@@ -52,7 +53,7 @@ export function compress(
       zlib.deflate(dataToBeCompressed, zlibOptions, callback as zlib.CompressCallback);
       break;
     default:
-      throw new Error(
+      throw new MongoDriverError(
         'Attempt to compress message using unknown compressor "' +
```
[the remark this answers is not in the document]
Comment (dariakp) [reply src/cmap/wire_protocol/compression.ts:57]: resolving for intern work

File: src/cmap/wire_protocol/compression.ts [hunk 10]
```diff
@@ -67,7 +68,7 @@ export function decompress(
   callback: Callback<Buffer>
 ): void {
   if (compressorID < 0 || compressorID > Math.max(2)) {
-    throw new Error(
+    throw new MongoDriverError(
       `Server sent message compressed using an unsupported compressor.` +
```
[the remark this answers is not in the document]
Comment (dariakp) [reply src/cmap/wire_protocol/compression.ts:72]: resolving for intern work

File: src/gridfs-stream/download.ts (hunk 4 above)
Comment (dariakp) [reply src/gridfs-stream/download.ts:258]: resolving for intern work

File: src/bulk/common.ts (hunk 1 above)
Comment (dariakp) [reply src/bulk/common.ts:658]: NODE-3337

File: src/gridfs-stream/download.ts (hunk 3 above)
Comment (dariakp) [reply src/gridfs-stream/download.ts:319]: NODE-3338

File: src/collection.ts (hunk 6 above)
Comment (dariakp) [reply src/collection.ts:1401]: NODE-3339

**Review by emadum — COMMENTED** [review]
LGTM!
